### PR TITLE
Convert UserDAO to typescript & rename to UserDAL

### PR DESCRIPTION
**Pull request by rizwanmustafa**
To typescript

### CODE REVIEW

**Comment by rizwanmustafa** [comment]
@Bruception @Miodec A review please.

File: backend/dao/user.ts [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+  static async isNameAvailable(name): Promise<boolean> {
+    const nameDocs = await db
+      .collection<MonkeyTypes.User>("users")
+      .find({ name })
+      .collation({ locale: "en", strength: 1 })
+      .limit(1)
+      .toArray();
+    if (nameDocs.length !== 0) {
+      return false;
+    } else {
```
Comment (Bruception) [inline backend/dao/user.ts:89]: Useless else

File: backend/dao/user.ts [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+      if (pb.LbPersonalBests) {
+        await db
+          .collection<MonkeyTypes.User>("users")
+          .updateOne(
+            { uid },
+            { $set: { lbPersonalBests: pb.LbPersonalBests } }
+          );
+      }
+      return true;
+    } else {
```
Comment (Bruception) [inline backend/dao/user.ts:269]: Useless else

File: backend/dao/user.ts [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+      best60 = undefined;
+    }
+
+    if (best60 === undefined || wpm >= best60 - best60 * 0.25) {
+      //increment when no record found or wpm is within 25% of the record
+      return await db
+        .collection<MonkeyTypes.User>("users")
+        .updateOne({ uid }, { $inc: { bananas: 1 } });
+    } else {
```
Comment (Bruception) [inline backend/dao/user.ts:408]: Useless else

File: backend/dao/user.ts [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+      .collection<MonkeyTypes.User>("users")
+      .findOne({ uid });
+
+    if (!user) {
+      throw new MonkeyError(404, "User not found", "Get personal bests");
+    }
+
+    if (mode2) {
+      return user?.personalBests?.[mode]?.[mode2];
+    } else {
```
Comment (Bruception) [inline backend/dao/user.ts:524]: Useless else

File: backend/utils/pb.ts [hunk 5]
```diff
@@ -2,18 +2,19 @@ import _ from "lodash";
 
 interface CheckAndUpdatePbResult {
   isPb: boolean;
-  obj: object;
-  lbObj?: object;
+  personalBests: MonkeyTypes.PersonalBests;
+  LbPersonalBests?: MonkeyTypes.LbPersonalBests;
```
Comment (Bruception) [inline backend/utils/pb.ts:6]: Why is first letter capital?

File: backend/dao/user.ts [hunk 6]
```diff
@@ -0,0 +1,530 @@
+import _ from "lodash";
+import { isUsernameValid } from "../utils/validation";
+import { updateUserEmail } from "../utils/auth";
+import { checkAndUpdatePb } from "../utils/pb";
+import db from "../init/db";
+import MonkeyError from "../utils/error";
+import { DeleteResult, InsertOneResult, ObjectId, UpdateResult } from "mongodb";
+
+class UsersDAO {
```
Comment (Bruception) [inline backend/dao/user.ts:9]: Might as well remove class syntax while you are at it.

**Comment by rizwanmustafa** [comment]
@Miodec  @Bruception A final review please. Thanks

**Comment by rizwanmustafa** [comment]
@Miodec  @Bruception A final review please. Thanks

**Review by Bruception — APPROVED** [review]
Let’s move from dao to something else. Since DAO is not the pattern we are using.

**Review by Bruception — CHANGES_REQUESTED** [review]
.

File: backend/utils/pb.ts [hunk 7]
```diff
@@ -40,8 +41,8 @@ export function checkAndUpdatePb(
 
   return {
     isPb,
-    obj: userPb,
-    lbObj: lbPersonalBests,
+    personalBests: userPb,
+    lbPersonalBests: lbPersonalBests,
   };
```
[the remark this answers is not in the document]
Comment (rizwanmustafa) [reply backend/utils/pb.ts:46]: I replaced it through VScode so it should be fine.

**Review by Miodec — CHANGES_REQUESTED** [review]
Lets make a new folder `/dal/` and move the converted dao's there

**Review by Miodec — APPROVED** [review]

**Review by Miodec — APPROVED** [review]

**Pull request merged**